drivers/usb/udc/udc_stm32.c

```diff
@@ -0,0 +1,1156 @@
+/*
+ * Copyright (c) 2023 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file  udc_stm32.c
+ * @brief STM32 USB device controller (UDC) driver
+ */
+
+#include <soc.h>
+#include <stm32_ll_bus.h>
+#include <stm32_ll_pwr.h>
+#include <stm32_ll_rcc.h>
+#include <stm32_ll_system.h>
+#include <string.h>
+#include <zephyr/usb/usb_device.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/pinctrl.h>
+#include "stm32_hsem.h"
+
+#include "udc_common.h"
+
+#include <zephyr/logging/log.h>
+#include <zephyr/irq.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(udc_stm32, CONFIG_UDC_DRIVER_LOG_LEVEL);
+
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs)
+#define DT_DRV_COMPAT st_stm32_otghs
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otgfs)
+#define DT_DRV_COMPAT st_stm32_otgfs
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_usb)
+#define DT_DRV_COMPAT st_stm32_usb
+#endif
```

Comment:
Will it work with multiple driver instances, FS + HS?

Reply:
The way I remember it, the HS and FS are mutually exclusive at the hardware level. Do I recall correctly or can these actually be used simultaneously?

Reply:
It appears to be possible, see [stm32f7-usbfs-and-usbhs-together](https://community.st.com/s/question/0D50X00009XkZ5kSAF/stm32f7-usbfs-and-usbhs-together) This being said, it isn't possible in current driver version so this is not a new limitation which is being added here. Driver could be converted later to multiple, heterogeneous instances.

Reply:
Okay. Just for notice, the native implementation for DWC2 controller will support multiple FS/HS instances combination.